spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/cache/Cache2kCacheConfiguration.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2012-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.boot.autoconfigure.cache;
+
+import java.util.Collection;
+
+import org.cache2k.Cache2kBuilder;
+import org.cache2k.extra.spring.SpringCache2kCacheManager;
+
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.cache.CacheManager;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Conditional;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.util.CollectionUtils;
+
+/**
+ * Support for cache2k auto configuration.
+ *
+ * @author Jens Wilke
+ */
+@Configuration(proxyBeanMethods = false)
+@ConditionalOnClass({ Cache2kBuilder.class, SpringCache2kCacheManager.class })
+@ConditionalOnMissingBean(CacheManager.class)
+@Conditional({ CacheCondition.class })
+class Cache2kCacheConfiguration {
+
+	@Bean
+	SpringCache2kCacheManager cacheManager(CacheProperties cacheProperties, CacheManagerCustomizers customizers) {
+		String managerName = cacheProperties.getCache2k().getManagerName();
+		SpringCache2kCacheManager cacheManager = (managerName != null) ? new SpringCache2kCacheManager(managerName)
+				: new SpringCache2kCacheManager();
+		customizers.customize(cacheManager);
+		addListedButUnknownCaches(cacheManager, cacheProperties.getCacheNames());
+		return cacheManager;
+	}
+
+	/**
+	 * Add caches yet unknown that are configured via the {@code CacheProperties}. Don't
+	 * add those before customization to be able the add the cache with programmatic
+	 * configuration via customization. These caches are created with default parameters
+	 * provided either via the customizer or the XML configuration
+	 * @param cacheManager the cache manger to add the caches
+	 * @param cacheNames the cache names
+	 */
+	private void addListedButUnknownCaches(SpringCache2kCacheManager cacheManager, Collection<String> cacheNames) {
```

Comment:
This is at odd with all the other implementations. Customizers run after cache names have been set. Properties or bean defined in the context are used to customize how caches are created by default. Because cache names are properties-based as well, it is expected to be processed before the customizers run and I am not keen to change that order for Cache2K only.

Cache2K has a `defaultSetup` function but I don't think exposing that as a bean is a good idea. I am tempted to expose this:

```java
public interface Cache2kBuilderCustomizer {

    void customize(Cache2kBuilder<?, ?> builder);

}
```

And call `defaultSetup` with a function that would invoke those. This makes `CacheManagerCustomizer` less useful but that is already the case for anything that changes the default setup.

Reply:
Good point.

Checking the other implementations more thoroughly I found another issue: E.g. the implementations Caffeine, JCache, SimpleCacheConfiguration allow no dynamic creation if the list of names is provided. That seems to be a standard and should be added also.

It might also be a good idea to pass the class loader, like in the JCache scenario. Cache2k allows setting the classloader and uses it currently when creating class instances from the external configuration. In the future more serialization/deserialization will happen.

W.r.t. to the default setup: The idea `Cache2kBuilderCustomizer` does the job simple and well, however: It is introducing another new concept and it is not inherently clear that this is actually the default setup. Providing the configuration bean `Cache2kConfig` seems to be a better fit. Ideally, this could be simply added to the cache properties. I will double check, whether this works.